In [161]:
import pandas as pd
import numpy as np

In [162]:
data = pd.read_csv("diabetes.csv")

In [163]:
data.corr(numeric_only=True)['Outcome']

Pregnancies                 0.221898
Glucose                     0.466581
BloodPressure               0.065068
SkinThickness               0.074752
Insulin                     0.130548
BMI                         0.292695
DiabetesPedigreeFunction    0.173844
Age                         0.238356
Outcome                     1.000000
Name: Outcome, dtype: float64

____
#### ***Here `BloodPressure` and `SkinThickness` are not so much matter for output but we keep it our dataset, bcz we are doing hyperparemeter tunning***
____
<table border="1" cellspacing="0" cellpadding="5">
  <tr>
    <th>Feature</th>
    <th>Value</th>
  </tr>
  <tr>
    <td>Pregnancies</td>
    <td>0.221898</td>
  </tr>
  <tr>
    <td>Glucose</td>
    <td>0.466581</td>
  </tr>
  <tr>
    <td>BloodPressure</td>
    <td>0.065068</td>
  </tr>
  <tr>
    <td>SkinThickness</td>
    <td>0.074752</td>
  </tr>
  <tr>
    <td>Insulin</td>
    <td>0.130548</td>
  </tr>
  <tr>
    <td>BMI</td>
    <td>0.292695</td>
  </tr>
  <tr>
    <td>DiabetesPedigreeFunction</td>
    <td>0.173844</td>
  </tr>
  <tr>
    <td>Age</td>
    <td>0.238356</td>
  </tr>
  <tr>
    <td>Outcome</td>
    <td>1.000000</td>
  </tr>
</table>


In [164]:
x = data.iloc[:, :-1].values
y = data.iloc[:, -1].values

In [165]:
# scaling
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaled_x = scaler.fit_transform(x)

In [166]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(scaled_x, y, test_size=0.2, random_state=42)

In [217]:
from tensorflow.keras.layers import Dense, Dropout
from tensorflow.keras.models import Sequential

In [168]:
model = Sequential()
model.add(
    Dense(32, input_dim=8, activation='relu')
)
model.add(
    Dense(1, activation='sigmoid')
)

c:\Users\user\AppData\Local\Programs\Python\Python313\Lib\site-packages\keras\src\layers\core\dense.py:92: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [169]:
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

In [170]:
history = model.fit(x_train, y_train, batch_size=32, verbose=1, epochs=100, validation_data=(x_test, y_test))

Epoch 1/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 1s 10ms/step - accuracy: 0.6319 - loss: 0.6494 - val_accuracy: 0.7338 - val_loss: 0.5855
Epoch 2/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.6938 - loss: 0.6007 - val_accuracy: 0.7922 - val_loss: 0.5508
Epoch 3/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7117 - loss: 0.5654 - val_accuracy: 0.7727 - val_loss: 0.5259
Epoch 4/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7362 - loss: 0.5400 - val_accuracy: 0.7662 - val_loss: 0.5099
Epoch 5/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7476 - loss: 0.5210 - val_accuracy: 0.7792 - val_loss: 0.4985
Epoch 6/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7573 - loss: 0.5063 - val_accuracy: 0.7922 - val_loss: 0.4906
Epoch 7/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7638 - loss: 0.4954 - val_accuracy: 0.7987 - val_loss: 0.4866
Epoch 8/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7704 - loss: 0.4871 - val_accuracy: 0.7987 - 

<div>
  <h2>🌱 What is Keras Tuner?</h2>

  <p><strong>Keras Tuner</strong> is a tool for automatic hyperparameter tuning in deep learning.</p>

  <p>When you build a neural network, you have to make choices like:</p>

  <ul>
    <li>How many layers to use?</li>
    <li>How many neurons per layer?</li>
    <li>Which activation function?</li>
    <li>What learning rate?</li>
    <li>Which optimizer?</li>
  </ul>

  <p>These choices are called <strong>hyperparameters</strong>. They are not learned by the model — you set them before training.</p>
</div>


In [171]:
import keras_tuner as kt

In [172]:
# def build_model(hp):
#     model = Sequential([
#         Dense(32, activation='relu', input_dim=8),
#         Dense(1, activation='sigmoid')
#     ])

#     optimizer = hp.Choice('optimizer', values=['adam', 'sgd', 'rmsprop', 'adadelta'])
#     model.compile(optimizer=optimizer, loss='binary_crossentropy', metrics=['accuracy'])
#     return model

In [173]:
# # creating tuner object
# tuner = kt.RandomSearch(build_model,
#                         objective='val_accuracy',
#                         max_trials=5
#                         )

In [174]:
# tuner.search(x_train, y_train, epochs=5, validation_data=(x_test, y_test))

In [175]:
# tuner.get_best_hyperparameters()[0].values

In [176]:
# tuner.results_summary()


In [177]:
# # Now you can get the best model as well
# model = tuner.get_best_models(num_models=1)[0]

In [178]:
# model.summary()

In [179]:
# model.fit(x_train, y_train, batch_size=32, epochs=100, initial_epoch=6, validation_data=(x_test, y_test))

In [185]:
def build_model(hp):
    neurons = hp.Choice('units', [8, 32, 64, 128])
    # neurons = hp.Int('units', min_value=8, max_value=128, step=8)
    model = Sequential([
        Dense(neurons, input_dim=8, activation='relu'),
        Dense(1, activation='sigmoid'),
    ])
    model.compile(optimizer='adam', metrics=['accuracy'], loss='binary_crossentropy')
    return model
# def build_model(hp):
#   model = Sequential()
#   model.add(Dense(
#       hp.Choice('units', [8, 16, 32]),
#       activation='relu'))
#   model.add(Dense(1, activation='sigmoid'))
#   model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
#   return model

In [186]:
tune = kt.RandomSearch(
    build_model, 
    objective='val_accuracy',
    max_trials=5
)

In [187]:
tune.search(x_train, y_train, epochs=5, validation_data=(x_test, y_test))

Trial 4 Complete [00h 00m 02s]
val_accuracy: 0.7532467246055603

Best val_accuracy So Far: 0.7792207598686218
Total elapsed time: 00h 00m 08s


In [188]:
tune.results_summary()

Results summary
Results in .\untitled_project
Showing 10 best trials
Objective(name="val_accuracy", direction="max")

Trial 2 summary
Hyperparameters:
units: 32
Score: 0.7792207598686218

Trial 3 summary
Hyperparameters:
units: 128
Score: 0.7532467246055603

Trial 1 summary
Hyperparameters:
units: 64
Score: 0.7337662577629089

Trial 0 summary
Hyperparameters:
units: 8
Score: 0.6168830990791321


In [193]:
tune.get_best_hyperparameters()[0].values

{'units': 32}

In [196]:
# Get all trials from the tuner
trials = tune.oracle.trials

# Convert to a DataFrame
results = []
for trial_id, trial in trials.items():
    results.append({
        "trial_id": trial_id,
        "hyperparameters": trial.hyperparameters.values,
        "score": trial.score
    })

df = pd.DataFrame(results)

In [197]:
df

trial_id hyperparameters     score
0        0    {'units': 8}  0.616883
1        1   {'units': 64}  0.733766
2        2   {'units': 32}  0.779221
3        3  {'units': 128}  0.753247

Number of Hidden Layers

In [199]:
def build_model(hp):
    model = Sequential()
    model.add(Dense(8, input_dim=8, activation='relu'))

    # for checking number of hidden layers
    for i in range(hp.Int('num_layers', min_value=1, max_value=10)):
        model.add(Dense(8, activation='relu'))

    model.add(Dense(1, activation='sigmoid'))
    model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
    return model


In [200]:
tuner = kt.RandomSearch(
    build_model, 
    objective='val_accuracy',
    max_trials=5,
    directory='tuners',
    project_name='num_layers'
)

c:\Users\user\AppData\Local\Programs\Python\Python313\Lib\site-packages\keras\src\layers\core\dense.py:92: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [201]:
tuner.search(x_train, y_train, epochs=5, validation_data=(x_test, y_test))

Trial 5 Complete [00h 00m 04s]
val_accuracy: 0.6428571343421936

Best val_accuracy So Far: 0.701298713684082
Total elapsed time: 00h 00m 17s


In [205]:
tuner.get_best_hyperparameters()[0].values

{'num_layers': 9}

In [204]:
best_model = tuner.get_best_models(num_models=1)[0]

c:\Users\user\AppData\Local\Programs\Python\Python313\Lib\site-packages\keras\src\layers\core\dense.py:92: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
c:\Users\user\AppData\Local\Programs\Python\Python313\Lib\site-packages\keras\src\saving\saving_lib.py:797: UserWarning: Skipping variable loading for optimizer 'adam', because it has 2 variables whereas the saved optimizer has 46 variables. 
  saveable.load_own_variables(weights_store.get(inner_path))


In [207]:
best_model.fit(x_train, y_train, epochs=200, initial_epoch=6, validation_data=(x_test, y_test))

Epoch 7/200
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.8485 - loss: 0.3697 - val_accuracy: 0.7273 - val_loss: 0.5652
Epoch 8/200
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.8567 - loss: 0.3684 - val_accuracy: 0.7403 - val_loss: 0.5675
Epoch 9/200
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.8534 - loss: 0.3675 - val_accuracy: 0.7273 - val_loss: 0.5636
Epoch 10/200
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.8518 - loss: 0.3697 - val_accuracy: 0.7338 - val_loss: 0.5674
Epoch 11/200
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.8518 - loss: 0.3696 - val_accuracy: 0.7143 - val_loss: 0.5850
Epoch 12/200
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.8404 - loss: 0.3758 - val_accuracy: 0.7208 - val_loss: 0.5603
Epoch 13/200
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.8534 - loss: 0.3707 - val_accuracy: 0.7338 - val_loss: 0.5678
Epoch 14/200
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.8453 - loss: 0.3832 - val_accuracy: 0.740

For all types of tunning

In [218]:
def build_model(hp):
    model = Sequential()
    
    # 1. How many hidden layers to use?
    for i in range(hp.Int('num_layers', min_value=1, max_value=4)):
        # 2. How many neurons per layer?
        model.add(Dense(
            units=hp.Int(f'units_{i}', min_value=16, max_value=128, step=16),
            activation=hp.Choice('activation', values=['relu', 'tanh', 'sigmoid'])
        ))
        # dropouts
        model.add(Dropout(rate=hp.Float(f'dropout_{i}', min_value=0.0, max_value=0.5, step=0.1)))
    
    # Output layer (binary classification case)
    model.add(Dense(1, activation='sigmoid'))
    
    # 3. What learning rate? + 4. Which optimizer?
    optimizer_name = hp.Choice('optimizer', ['adam', 'sgd', 'rmsprop', 'adadelta'])
    
    model.compile(optimizer=optimizer_name, 
                  loss='binary_crossentropy', 
                  metrics=['accuracy'])
    return model


In [219]:
tuner = kt.RandomSearch(
    build_model, 
    objective='val_accuracy',
    max_trials=5,
    directory='tuners',
    project_name='all_params_final'
)

In [220]:
tuner.search(x_train, y_train, epochs=5, validation_data=(x_test, y_test))

Trial 5 Complete [00h 00m 03s]
val_accuracy: 0.6428571343421936

Best val_accuracy So Far: 0.7727272510528564
Total elapsed time: 00h 00m 14s


In [221]:
tuner.get_best_hyperparameters()[0].values

{'num_layers': 1,
 'units_0': 48,
 'activation': 'relu',
 'dropout_0': 0.0,
 'optimizer': 'adam',
 'units_1': 64,
 'dropout_1': 0.4,
 'units_2': 32,
 'dropout_2': 0.30000000000000004}